spinedb_api/db_mapping_base.py

```diff
@@ -440,7 +484,10 @@ def checked_item_and_error(self, item, for_update=False):
         if error:
             return None, error
         valid_types = (type(None),) if for_update else ()
+        self.check_fields_for_addition(candidate_item)
         self.check_fields(candidate_item._asdict(), valid_types=valid_types)
+        if not for_update:
+            candidate_item.convert_dicts_db_ids_to_item_ids(self._item_type, candidate_item, self._db_map)
```

Comment:
I have to say I am not a big fan of replacing all references to DB ids with the in-mem (negative) ids at this point. Previously we'd keep the DB ids untouched and do the translation via `_MappedTable.get`. That was a bit more terse and seemed to work ok, but maybe I'm missing something?

Reply:
I cannot say I am a big fan of the temporary ids either but I spent a lot of time trying to come up with another solution that would fix refreshing and I always hit a wall. Using database ids in DatabaseMapping just does not work on refresh if the ids have been messed up by another database user, e.g. a script.

I hope the following example will show what problem the temporary ids solve.

Say, we have a database with single entity class, "Widget" with id 1. We have a `DatabaseMapping` object called `db_map` and we read the entity class into memory by calling `fetch_all()`. It gets a temporary id -1.

Then, user executes a script in a another process. The script purges all entity classes and adds two new entity classes, "Gadget" and "Widget". Gadget gets id 1 and Widget gets id 2. The important part here is that Widget's database id was changed from 1 to 2.

Handling Widget's new id would not have worked with the previous implementation that was using real database ids because those ids could not change on-the-fly without us implementing a complex notification mechanism.

When we call `db_map.fetch_all()` again with the current implementation, we keep Widget's temporary id as is. This means that Users of `db_map` can keep referring to Widget by -1 and not worry about its new database id.

So, ultimately we want to disconnect the in-memory ids from the real database ids to be able to deal with changes to the database by "outside forces".

Reply:
Thanks for the explanation! I have a couple of questions:
- So what happens when Widget shows up with a different id in the second fetch. I guess we assume it is the same as the Widget from the previous fetch because the unique keys are the same. So we don't add it back. But if the second fetch also brings some entities of the Widget class, their class_id would be 2. Are we able to recognize that they belong to the Widget class, even if the class_id is the one from the second fetch?
- Say the temp ids work, my comment was more about implementation. Instead of replacing the db ids with item ids in each item, I was suggesting to keep the db ids and then find the referred items via the id map. For example, say an entity item has class_id=5 and we want the class_name, we'd look for the item_id corresponding to the db_id 5 and then find the entity_class.

Reply:
Trying to answer my own question... when the 'same' items start showing up with different ids in subsequent fetches,
it looks like we don't update the IdMap - we always trust the id of the first 'version' of the item.
That means all items that reference the 'second' version by id will have trouble finding the item.
What we would need to do maybe is build a map from internal in-mem id to a *list* of all the db ids it represents,
for all the times the same item shows up with a different id. But that is getting increasingly complicated and makes me wonder if we really want to support dealing with external changes at this point where our priority should be to try and stabilize?

Maybe we should satisfy ourselves with something very raw like checking if the db has new commits after we started
fetching this DatabaseMapping, and in that case refuse to fetch again as it can lead to chaos?
What do you think?

Reply:
I think that issue can be fixed by simply updating the IdMap when an item is refetched.

Refusing fetching crossed my mind as well but I think that is a no-go. It would be a mess if a user had Database editor open with pending changes and they executed Importer in the background.

Reply:
> I think that issue can be fixed by simply updating the IdMap when an item is refetched

Cool, are you thinking about keeping a mapping from temp id to a set of all db ids it represents?

We need to weigh the benefits versus complexity in my opinion. My impression is we are adding an entire new layer of complexity to the code to support a very specific use case. I'm just a bit worried about stability. It would be nice to have something we can trust and merge into master soon. These changes might interfere with that goal. I would still like to consider reverting and keeping this as a PR - but maybe I am exaggerating and this is completely fine, provided we can fix the bugs with fetching references, and are also able to fix the issue I just signaled above. But maybe we will keep finding issues.

I am sorry for not being able to review this earlier.

Reply:
> I would still like to consider reverting and keeping this as a PR 

We should definitely revert if we do not feel comfortable with the changes. I am fine with the current complexity but a bit afraid at the same time that something comes up that this approach cannot solve without extensive and complex changes.

I will open PRs that revert the changes. It is better to work on this in a development branch until we gain more trust and consensus on the in-memory id.

> I am sorry for not being able to review this earlier.

No worries. Sorry for dragging you out and forcing you to review in such a drastic way :)